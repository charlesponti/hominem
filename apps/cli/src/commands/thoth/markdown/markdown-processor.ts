import { logger } from '@ponti/utils/logger'
import type { TextAnalysis } from '@ponti/utils/nlp'
import { getDatesFromText } from '@ponti/utils/time'
import * as cheerio from 'cheerio'
import * as fs from 'node:fs/promises'
import rehypeStringify from 'rehype-stringify'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { unified } from 'unified'
import { detectTask, normalizeWhitespace, taskRegex } from './utils'

interface GetProcessedEntryParams {
  $: ReturnType<typeof cheerio.load>
  elem: ReturnType<ReturnType<typeof cheerio.load>>[0]
  entry: ProcessedMarkdownFileEntry
  section: string | null
  tag: string
}

export interface EntryContent {
  tag: string
  text: string
  section: string | null
  isTask?: boolean
  isComplete?: boolean
  textAnalysis?: TextAnalysis
  subentries?: EntryContent[]
}

export interface ProcessedMarkdownFileEntry {
  content: EntryContent[]
  date: string | undefined
  filename: string
  heading: string
  frontmatter?: Record<string, unknown>
}

export interface ProcessedMarkdownFile {
  entries: ProcessedMarkdownFileEntry[]
}

export class MarkdownProcessor {
  async processFileWithAst(filepath: string): Promise<ProcessedMarkdownFile> {
    // Check that the file exists
    if (!(await fs.stat(filepath)).isFile()) {
      throw new Error(`File not found: ${filepath}`)
    }

    const content = await fs.readFile(filepath, 'utf-8')
    const { result } = await this.convertMarkdownToJSON(content, filepath)
    return result
  }

  processFrontmatter(content: string): { content: string; frontmatter?: Record<string, unknown> } {
    // Find the frontmatter content at the start of the file
    const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---\n/)
    let frontmatter: Record<string, unknown> | undefined
    let processableContent = content

    if (frontmatterMatch) {
      try {
        // Parse the frontmatter content (assuming YAML format)
        const frontmatterContent = frontmatterMatch[1]
        // For now, we'll do a simple key-value extraction
        frontmatter = Object.fromEntries(
          frontmatterContent
            .split('\n')
            .filter((line) => line.includes(':'))
            .map((line) => {
              const [key, ...valueParts] = line.split(':')
              const value = valueParts.join(':').trim()
              // Remove quotes if present
              return [key.trim(), value.replace(/^['"]|['"]$/g, '')]
            })
        )
        // Remove the frontmatter from the content
        processableContent = content.slice(frontmatterMatch[0].length)
      } catch (error) {
        logger.warn('Failed to parse frontmatter:', error)
      }
    }

    return { content: processableContent, frontmatter }
  }

  async convertMarkdownToHTML(content: string) {
    const file = await unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeStringify)
      .process(content)

    return cheerio.load(String(file))
  }

  convertMarkdownToAST(content: string) {
    const processor = unified().use(remarkParse)
    return processor.parse(content)
  }

  // Helper method to create or retrieve an entry
  private ensureEntry(
    filename: string,
    heading: string | undefined,
    frontmatter?: Record<string, unknown>,
    entries: ProcessedMarkdownFileEntry[] = []
  ): ProcessedMarkdownFileEntry {
    // If no entry exists or we're at a new heading, create a new entry
    const entry: ProcessedMarkdownFileEntry = {
      date: undefined,
      filename,
      heading: heading || filename.split('/').pop() || '',
      content: [],
      frontmatter,
    }

    entries.push(entry)
    return entry
  }

  async convertMarkdownToJSON(
    markdownContent: string,
    filename: string
  ): Promise<{ result: ProcessedMarkdownFile; html: string }> {
    const result: ProcessedMarkdownFile = { entries: [] }

    // Extract frontmatter if present
    const { content: processableContent, frontmatter } = this.processFrontmatter(markdownContent)

    // Convert markdown to HTML for fancy querying
    const $ = await this.convertMarkdownToHTML(processableContent)

    let currentHeading: string | undefined
    let currentEntry: ProcessedMarkdownFileEntry | null = null

    const elements = $('body').children().toArray()
    for (const elem of elements) {
      const tag = elem.tagName.toLowerCase()
      const text = $(elem).text().trim()
      if (!text) continue

      const { taskText } = detectTask(text)

      const processedText =
        taskText ||
        text
          .replace(/^(-|\*)\s+/, '')
          .replace(taskRegex, '')
          .trim()

      if (/^h[1-6]$/.test(tag)) {
        // New entry on headings
        currentHeading = text
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
        const { dates, fullDate } = getDatesFromText(text)
        currentEntry.date = dates.length > 0 ? dates[0].start.split('T')[0] : fullDate
      }

      if (tag === 'paragraph') {
        const previousContent = this.getPreviousEntry(currentEntry)

        if (previousContent && previousContent.tag === 'paragraph') {
          // Add to previous paragraph
          previousContent.text += `\n ${processedText}`
        }

        continue
      }

      if (!currentEntry) {
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
      }

      if (tag === 'ul' || tag === 'ol') {
        const listItems = $(elem).find('> li').toArray()
        for (const li of listItems) {
          await this.processListItem({
            $,
            elem: li,
            entry: currentEntry,
            section: currentHeading ? currentHeading.toLowerCase() : null,
            tag,
          })
        }

        continue
      }

      const processedContent = await this.getProcessedEntry({
        $,
        elem,
        entry: currentEntry,
        section: currentHeading ? currentHeading.toLowerCase() : null,
        tag,
      })
      if (processedContent) {
        currentEntry.content.push(processedContent)
      }
    }

    if (result.entries.length === 0 && frontmatter) {
      this.ensureEntry(filename, undefined, frontmatter, result.entries)
    }

    return { result, html: $.html() }
  }

  async getProcessedEntry(params: GetProcessedEntryParams): Promise<EntryContent | undefined> {
    const { $, elem, section, tag } = params
    const text = $(elem).contents().first().text().trim()
    if (!text) return
    const normalizedText = normalizeWhitespace(text)
    const { isTask, isComplete, taskText } = detectTask(normalizedText)

    return {
      tag,
      text: taskText || normalizedText,
      section,
      isTask,
      isComplete,
      subentries: [],
    }
  }

  async processListItem(params: GetProcessedEntryParams) {
    const { $, elem, entry } = params
    const processedContent = await this.getProcessedEntry(params)
    if (!processedContent) return

    entry.content.push(processedContent)

    // Refactored nested lists recursion
    const subentries = await this.processNestedLists($, elem, entry, processedContent.tag)
    if (subentries.length > 0) {
      processedContent.subentries = subentries
    }
  }

  private async processNestedLists(
    $: ReturnType<typeof cheerio.load>,
    liElem: ReturnType<ReturnType<typeof cheerio.load>>[0],
    entry: ProcessedMarkdownFileEntry,
    tag: string
  ): Promise<EntryContent[]> {
    const subentries: EntryContent[] = []

    // Process immediate nested lists (ul or ol) inside the current li element
    const nestedLists = $(liElem).find('> ul, > ol').toArray()
    for (const nestedList of nestedLists) {
      // Use the text of the current li as section
      const parentText = $(liElem).contents().first().text().trim()
      const nestedListItems = $(nestedList).find('> li').toArray()
      for (const nestedListItem of nestedListItems) {
        const nestedContent = await this.getProcessedEntry({
          $,
          elem: nestedListItem,
          entry,
          section: parentText,
          tag,
        })
        if (nestedContent) {
          // Process deeper nested lists recursively
          nestedContent.subentries = await this.processNestedLists($, nestedListItem, entry, tag)
          subentries.push(nestedContent)
        }
      }
    }
    return subentries
  }

  getPreviousEntry(entry: ProcessedMarkdownFileEntry | null): EntryContent | undefined {
    if (!entry) return undefined
    return entry.content[entry.content.length - 1]
  }
}
