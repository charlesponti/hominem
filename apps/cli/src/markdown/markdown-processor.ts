import * as mdast from 'mdast-util-to-string'
import remarkParse from 'remark-parse'
import { unified } from 'unified'
import type { Node } from 'unist'
import type { Note, ProcessedContent } from './types'
import { getDateFromText, sanitizeText } from './utils'

interface MarkdownNode extends Node {
  type: string
  children?: MarkdownNode[]
  value?: string
  depth?: number
}

export class MarkdownProcessor {
  private currentHeading: string | undefined
  private content: ProcessedContent

  constructor() {
    this.content = {
      headings: [],
      paragraphs: [],
      bulletPoints: [],
      others: [],
    }
  }

  async processFile(filepath: string, content: string): Promise<ProcessedContent> {
    const tree = await unified().use(remarkParse).parse(content)

    this.traverseNodes(tree as MarkdownNode, filepath)
    return this.content
  }

  private traverseNodes(node: MarkdownNode, filename: string): void {
    if (node.type === 'heading') {
      const text = mdast.toString(node)
      this.currentHeading = sanitizeText(text)
      this.content.headings.push(this.currentHeading)
    }

    if (node.type === 'paragraph') {
      const text = mdast.toString(node)
      const { fullDate } = getDateFromText(text)

      this.content.paragraphs.push({
        file: filename,
        heading: this.currentHeading,
        text: sanitizeText(text),
        tag: 'paragraph',
        date: fullDate,
      })
    }

    if (node.type === 'list') {
      for (const item of node.children as MarkdownNode[]) {
        if (item.type === 'listItem') {
          const text = mdast.toString(item)
          const { fullDate } = getDateFromText(text)

          this.content.bulletPoints.push({
            file: filename,
            heading: this.currentHeading,
            text: sanitizeText(text),
            tag: 'bullet_point',
            date: fullDate,
          })
        }
      }
    }

    // Traverse children
    if (node.children) {
      for (const child of node.children) {
        this.traverseNodes(child, filename)
      }
    }
  }
}
