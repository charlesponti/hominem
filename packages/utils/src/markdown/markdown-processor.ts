import { getDatesFromText } from '@/time'
import * as cheerio from 'cheerio'
import {
  MarkdownTextSplitter,
  type RecursiveCharacterTextSplitterParams,
} from 'langchain/text_splitter'
import rehypeStringify from 'rehype-stringify'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { unified } from 'unified'
import type { TextAnalysis } from '../schemas'
import { extractMetadata, type Metadata } from './metadata.schema'
import { detectTask, normalizeWhitespace, taskRegex } from './utils'

export interface EntryContent {
  tag: string
  text: string
  section: string | null
  isTask?: boolean
  isComplete?: boolean
  textAnalysis?: TextAnalysis
  subentries?: EntryContent[]
}

interface GetProcessedEntryParams {
  $: ReturnType<typeof import('cheerio').load>
  elem: ReturnType<ReturnType<typeof import('cheerio').load>>[0]
  entry: {
    content: EntryContent[]
    date?: string
    filename: string
    heading: string
  }
  section: string | null
}

export interface ProcessedMarkdownFileEntry {
  content: EntryContent[]
  date?: string
  filename: string
  heading: string
  metadata?: Metadata
  frontmatter?: Record<string, unknown>
}

export interface ProcessedMarkdownFile {
  entries: ProcessedMarkdownFileEntry[]
  metadata?: Metadata
}

export class MarkdownProcessor {
  async getChunks(content: string, options?: Partial<RecursiveCharacterTextSplitterParams>) {
    const splitter = MarkdownTextSplitter.fromLanguage('markdown', {
      separators: ['#', '##', '###', '####', '#####', '######'],
      ...options,
    })
    const chunks = await splitter.splitText(content)
    return chunks
  }

  async processFileWithAst(content: string, filename: string): Promise<ProcessedMarkdownFile> {
    const { result } = await this.convertMarkdownToJSON(content, filename)

    return result
  }

  processFrontmatter(content: string): {
    content: string
    metadata: Metadata
    frontmatter?: Record<string, unknown>
  } {
    // Find the frontmatter content at the start of the file
    const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---\n/)
    let frontmatter: Record<string, unknown> | undefined
    let processableContent = content
    let metadata: Metadata = {}

    if (frontmatterMatch) {
      // Parse the frontmatter content (assuming YAML format)
      const frontmatterContent = frontmatterMatch[1]

      if (!frontmatterContent) {
        return { content, metadata, frontmatter: undefined }
      }

      // Extract key-value pairs
      frontmatter = Object.fromEntries(
        frontmatterContent
          .split('\n')
          .filter((line) => line.includes(':'))
          .map((line) => {
            const [key, ...valueParts] = line.split(':')
            const value = valueParts.join(':').trim()
            if (!key) return [value]
            return [key.trim(), value.replace(/^['"]|['"]$/g, '')]
          })
      )

      // Validate and extract metadata
      if (frontmatter) {
        metadata = extractMetadata(frontmatter)
      }

      // Remove the frontmatter from the content
      processableContent = content.slice(frontmatterMatch[0].length)
    }

    return { content: processableContent, metadata, frontmatter }
  }

  async convertMarkdownToHTML(content: string) {
    const file = await unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeStringify)
      .process(content)

    return cheerio.load(String(file))
  }

  convertMarkdownToAST(content: string) {
    const processor = unified().use(remarkParse)
    return processor.parse(content)
  }

  // Helper method to create or retrieve an entry
  private ensureEntry(
    filename: string,
    heading?: string,
    frontmatter?: Record<string, unknown>,
    entries: ProcessedMarkdownFileEntry[] = []
  ): ProcessedMarkdownFileEntry {
    // If no entry exists or we're at a new heading, create a new entry
    const entry: ProcessedMarkdownFileEntry = {
      date: undefined,
      filename,
      heading: heading || filename.split('/').pop() || '',
      content: [],
      frontmatter,
    }

    entries.push(entry)
    return entry
  }

  private async calculateReadingMetrics(
    content: string
  ): Promise<{ wordCount: number; readingTime: number }> {
    const words = content.trim().split(/\s+/).length
    // Average reading speed: 200-250 words per minute
    const readingTime = Math.ceil(words / 200)

    return { wordCount: words, readingTime }
  }

  async convertMarkdownToJSON(
    markdownContent: string,
    filename: string
  ): Promise<{ result: ProcessedMarkdownFile; html: string }> {
    // Process frontmatter and extract metadata
    const {
      content: processableContent,
      metadata,
      frontmatter,
    } = this.processFrontmatter(markdownContent)

    // Calculate reading metrics
    const { wordCount, readingTime } = await this.calculateReadingMetrics(processableContent)
    metadata.wordCount = wordCount
    metadata.readingTime = readingTime

    // Convert markdown to HTML for processing
    const $ = await this.convertMarkdownToHTML(processableContent)

    const result: ProcessedMarkdownFile = {
      entries: [],
      metadata,
    }

    let currentHeading: string | undefined
    let currentEntry: ProcessedMarkdownFileEntry | null = null

    const elements = $('body').children().toArray()

    for (const elem of elements) {
      const tag = elem.tagName.toLowerCase()
      const text = $(elem).text().trim()
      if (!text) continue

      const { taskText } = detectTask(text)

      const processedText =
        taskText ||
        text
          .replace(/^(-|\*)\s+/, '')
          .replace(taskRegex, '')
          .trim()

      if (/^h[1-6]$/.test(tag)) {
        currentHeading = text
        // New entry on headings
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
        const { dates, fullDate } = getDatesFromText(text)
        if (dates.length > 0 && dates[0]) {
          currentEntry.date = dates[0].start.split('T')[0]
        } else if (fullDate) {
          currentEntry.date = fullDate.split('T')[0]
        }
      }

      if (tag === 'paragraph') {
        const previousContent = this.getPreviousEntry(currentEntry)

        // If the previous entry was a paragraph, add to previous paragraph
        if (previousContent && previousContent.tag === 'paragraph') {
          previousContent.text += `\n ${processedText}`
        }

        continue
      }

      if (!currentEntry) {
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
      }

      const processedContent = await this.getProcessedEntry({
        $,
        elem,
        entry: currentEntry,
        section: currentHeading ? currentHeading.toLowerCase() : null,
      })

      if (processedContent) {
        currentEntry.content.push(processedContent)
      }

      if (tag === 'ul' || tag === 'ol') {
        const listItems = $(elem).find('> li').toArray()

        for (const li of listItems) {
          const processedContent = await this.getProcessedEntry({
            $,
            elem: li,
            entry: currentEntry,
            section: currentHeading ? currentHeading.toLowerCase() : null,
          })
          if (!processedContent) {
            console.error('No processed content found for list item', tag, processedText)
            continue
          }

          currentEntry.content.push(processedContent)

          processedContent.subentries = await this.processNestedLists(
            $,
            li,
            currentEntry,
            processedContent.tag
          )
        }
      }
    }

    if (result.entries.length === 0 && frontmatter) {
      this.ensureEntry(filename, undefined, frontmatter, result.entries)
    }

    return { result, html: $.html() }
  }

  async getProcessedEntry(params: GetProcessedEntryParams): Promise<EntryContent | undefined> {
    const { $, elem, section } = params
    const text = $(elem).contents().first().text().trim()
    if (!text) return
    const normalizedText = normalizeWhitespace(text)
    const { isTask, isComplete, taskText } = detectTask(normalizedText)

    return {
      tag: elem.type,
      text: taskText || normalizedText,
      section,
      isTask,
      isComplete,
      subentries: [],
    }
  }

  private async processNestedLists(
    $: ReturnType<typeof cheerio.load>,
    liElem: ReturnType<ReturnType<typeof cheerio.load>>[0],
    entry: ProcessedMarkdownFileEntry,
    tag: string
  ): Promise<EntryContent[]> {
    const subentries: EntryContent[] = []

    // Process immediate nested lists (ul or ol) inside the current li element
    const nestedLists = $(liElem).find('> ul, > ol').toArray()
    for (const nestedList of nestedLists) {
      // Use the text of the current li as section
      const parentText = $(liElem).contents().first().text().trim()
      const nestedListItems = $(nestedList).find('> li').toArray()
      for (const nestedListItem of nestedListItems) {
        const nestedContent = await this.getProcessedEntry({
          $,
          elem: nestedListItem,
          entry,
          section: parentText,
        })
        if (nestedContent) {
          // Process deeper nested lists recursively
          nestedContent.subentries = await this.processNestedLists($, nestedListItem, entry, tag)
          subentries.push(nestedContent)
        }
      }
    }
    return subentries
  }

  getPreviousEntry(entry: ProcessedMarkdownFileEntry | null): EntryContent | undefined {
    if (!entry) return undefined
    return entry.content[entry.content.length - 1]
  }
}
