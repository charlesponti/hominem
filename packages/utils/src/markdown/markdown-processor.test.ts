import { beforeEach, describe, expect, it, vi } from 'vitest'
import { MarkdownProcessor } from './markdown-processor'

// Mock the time module
vi.mock('../time.js', () => {
  return {
    getDatesFromText: () => ({
      dates: [],
      fullDate: undefined,
      year: undefined,
    }),
  }
})

describe('MarkdownProcessor', () => {
  let processor: MarkdownProcessor

  beforeEach(() => {
    // Reset any mocks before each test
    vi.clearAllMocks()
    processor = new MarkdownProcessor()
  })

  describe('processFrontmatter', () => {
    it('should extract frontmatter from markdown content', () => {
      const content = `---
title: Test Document
date: 2023-05-01
tags: test, markdown
---

# Heading 1

Some content here.`

      const result = processor.processFrontmatter(content)

      expect(result.frontmatter).toEqual({
        title: 'Test Document',
        date: '2023-05-01',
        tags: 'test, markdown',
      })

      expect(result.content).toContain('# Heading 1')
    })

    it('should handle markdown without frontmatter', () => {
      const content = '# Heading 1\n\nSome content here.'

      const result = processor.processFrontmatter(content)

      expect(result.frontmatter).toBeUndefined()
      expect(result.content).toBe(content)
    })
  })

  describe('convertMarkdownToJSON', () => {
    it('should process headings properly', async () => {
      const content =
        '# Heading 1\n\nParagraph under heading 1\n\n## Heading 2\n\nParagraph under heading 2'

      const { result } = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result).toMatchSnapshot()
    })

    it('should process lists correctly', async () => {
      const content = `# Shopping List
      
- Apples
- Bananas
- Milk`

      const { result } = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result).toMatchSnapshot()
    })

    it('should detect tasks and their completion status', async () => {
      const content = `# Tasks
      
- [ ] Incomplete task
- [x] Complete task`

      const { result } = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result).toMatchSnapshot()
    })

    it('should handle nested lists properly', async () => {
      const content = `# Nested List
      
- Fruits:
  - Apples
  - Bananas
- Vegetables:
  - Carrots
  - Broccoli`

      const { result } = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result).toMatchSnapshot()
    })

    it('should extract metadata from content', async () => {
      const content = `# Meeting Notes
      
Meeting with John Smith in New York about #project planning.`

      const { result } = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result).toMatchSnapshot()
    })

    it('should correctly process hierarchical lists with personal reflections', async () => {
      const markdown = `## personal
- **Perfectionism**
- **Overthinking, analysis paralysis, fear of failure**
  - decrease my playfulness.
  - decreases poor decision-making
  - decreases action-taking
  - reinforces personal beliefs instead of increasing objective truth
- **Fear of judgement**
  - desire to be seen as correct instead of desire to be effective
  - many judgments should be ignored because they lack sincerity`

      const { result } = await processor.convertMarkdownToJSON(markdown, 'personal-reflections.md')

      expect(result).toMatchSnapshot()
    })
  })

  describe('processFileWithAst', () => {
    it('should read file and process its content', async () => {
      const fileContent = '# Test Heading\n\nTest content'

      const result = await processor.processFileWithAst(fileContent, 'test.md')

      expect(result.entries.length).toBe(1)
      expect(result.entries[0]?.heading).toBe('Test Heading')
    })
  })
})
