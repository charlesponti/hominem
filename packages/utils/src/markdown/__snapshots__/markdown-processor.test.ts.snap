// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarkdownProcessor > convertMarkdownToJSON > should correctly process hierarchical lists with personal reflections 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [],
          "tag": "tag",
          "text": "personal",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [],
          "tag": "tag",
          "text": "Perfectionism",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "tag",
              "text": "decrease my playfulness.",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "tag",
              "text": "decreases poor decision-making",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "tag",
              "text": "decreases action-taking",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "tag",
              "text": "reinforces personal beliefs instead of increasing objective truth",
            },
          ],
          "tag": "tag",
          "text": "Overthinking, analysis paralysis, fear of failure",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fear of judgement",
              "subentries": [],
              "tag": "tag",
              "text": "desire to be seen as correct instead of desire to be effective",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fear of judgement",
              "subentries": [],
              "tag": "tag",
              "text": "many judgments should be ignored because they lack sincerity",
            },
          ],
          "tag": "tag",
          "text": "Fear of judgement",
        },
      ],
      "date": undefined,
      "filename": "personal-reflections.md",
      "frontmatter": undefined,
      "heading": "personal",
    },
  ],
  "metadata": {
    "readingTime": 1,
    "wordCount": 58,
  },
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should detect tasks and their completion status 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "tasks",
          "subentries": [],
          "tag": "tag",
          "text": "Tasks",
        },
        {
          "isComplete": false,
          "isTask": true,
          "section": "tasks",
          "subentries": [],
          "tag": "tag",
          "text": "Incomplete task",
        },
        {
          "isComplete": true,
          "isTask": true,
          "section": "tasks",
          "subentries": [],
          "tag": "tag",
          "text": "Complete task",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Tasks",
    },
  ],
  "metadata": {
    "readingTime": 1,
    "wordCount": 11,
  },
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should extract metadata from content 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "meeting notes",
          "subentries": [],
          "tag": "tag",
          "text": "Meeting Notes",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "meeting notes",
          "subentries": [],
          "tag": "tag",
          "text": "Meeting with John Smith in New York about #project planning.",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Meeting Notes",
    },
  ],
  "metadata": {
    "readingTime": 1,
    "wordCount": 13,
  },
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should handle nested lists properly 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "nested list",
          "subentries": [],
          "tag": "tag",
          "text": "Nested List",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "nested list",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fruits:",
              "subentries": [],
              "tag": "tag",
              "text": "Apples",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fruits:",
              "subentries": [],
              "tag": "tag",
              "text": "Bananas",
            },
          ],
          "tag": "tag",
          "text": "Fruits:",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "nested list",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Vegetables:",
              "subentries": [],
              "tag": "tag",
              "text": "Carrots",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Vegetables:",
              "subentries": [],
              "tag": "tag",
              "text": "Broccoli",
            },
          ],
          "tag": "tag",
          "text": "Vegetables:",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Nested List",
    },
  ],
  "metadata": {
    "readingTime": 1,
    "wordCount": 15,
  },
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should process headings properly 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 1",
          "subentries": [],
          "tag": "tag",
          "text": "Heading 1",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 1",
          "subentries": [],
          "tag": "tag",
          "text": "Paragraph under heading 1",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Heading 1",
    },
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 2",
          "subentries": [],
          "tag": "tag",
          "text": "Heading 2",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 2",
          "subentries": [],
          "tag": "tag",
          "text": "Paragraph under heading 2",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Heading 2",
    },
  ],
  "metadata": {
    "readingTime": 1,
    "wordCount": 14,
  },
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should process lists correctly 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "tag",
          "text": "Shopping List",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "tag",
          "text": "Apples",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "tag",
          "text": "Bananas",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "tag",
          "text": "Milk",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Shopping List",
    },
  ],
  "metadata": {
    "readingTime": 1,
    "wordCount": 9,
  },
}
`;
